### ISSUE TITLE
Request for comment on NoMSVC

### ISSUE THREAD

**Issue by JohnScience** [mcgoo/vcpkg-rs#46]
@mcgoo Hi, which kind of Error is represented by NoMVC and why exactly does it exist? Was it a workaround for a bug or limitation of `vcpkg`? If so, can you link the issue?

**Comment by waych** [comment]
NoMSVC is, as far as I can tell, a left over from the fact that this crate originally only supported building with MSVC.  Later platforms such as Linux and iOS were added later, but the code still reports "NotMSVC" if the compiler triplet setup isn't recognized.  This should probably be cleaned up to be NotYetSupportedCompiler or something similar.

**Comment by JohnScience** [comment]
It appears [on line 1346](https://github.com/mcgoo/vcpkg-rs/blob/713efe4300f75b964b228a599562f2c0ac06d512/src/lib.rs#L1346) in the [body of `msvc_target` function](https://github.com/mcgoo/vcpkg-rs/blob/713efe4300f75b964b228a599562f2c0ac06d512/src/lib.rs#L1311-L1418), which is used in the `impl` block of `Config`, [on line 863](https://github.com/mcgoo/vcpkg-rs/blob/713efe4300f75b964b228a599562f2c0ac06d512/src/lib.rs#L863) of [the body of `get_target_triplet` function](https://github.com/mcgoo/vcpkg-rs/blob/713efe4300f75b964b228a599562f2c0ac06d512/src/lib.rs#L858-L869).

My concern is that I can't compile for `arm64-android` target of `vcpkg` while using `x86_64-pc-windows-msvc`. [Issue in Rust-SDL2](https://github.com/Rust-SDL2/rust-sdl2/issues/1266).

<details>
  <summary>Output of <code>vcpkg help triplet</code></summary>
  
  ```
  Available architecture triplets:
vcpkg built-in triplets:
  arm-uwp
  arm64-windows
  x64-linux
  x64-osx
  x64-uwp
  x64-windows-static
  x64-windows
  x86-windows
VCPKG community triplets:
  arm-android
  arm-ios
  arm-linux
  arm-mingw-dynamic
  arm-mingw-static
  arm-neon-android
  arm-windows-static
  arm-windows
  arm64-android
  arm64-ios
  arm64-linux
  arm64-mingw-dynamic
  arm64-mingw-static
  arm64-osx-dynamic
  arm64-osx
  arm64-uwp
  arm64-windows-static-md
  arm64-windows-static
  armv6-android
  ppc64le-linux
  s390x-linux
  wasm32-emscripten
  x64-android
  x64-freebsd
  x64-ios
  x64-linux-dynamic
  x64-linux-release
  x64-mingw-dynamic
  x64-mingw-static
  x64-openbsd
  x64-osx-dynamic
  x64-osx-release
  x64-windows-release
  x64-windows-static-md
  x86-android
  x86-freebsd
  x86-ios
  x86-mingw-dynamic
  x86-mingw-static
  x86-uwp
  x86-windows-static-md
  x86-windows-static
  x86-windows-v120
  ```
  
</details>

<details>
  <summary>Output of <code>rustc -Vv</code></summary>
  
  ```
  rustc 1.64.0 (a55dd71d5 2022-09-19)
binary: rustc
commit-hash: a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52
commit-date: 2022-09-19
host: x86_64-pc-windows-msvc
release: 1.64.0
LLVM version: 14.0.6
  ```
  
</details>

I'm not sure I know `vcpkg` well enough to solve the problem properly. Can you assist me with it, please? What would it take to solve the issue?

@waych

**Comment by JohnScience** [comment]
I bogged down in the rewriting of `TargetTriple` part of the code. Making it right is difficult as hell. Kudos to the person who made it work good 'nuff.

**Comment by waych** [comment]
@JohnScience yeah.. I took a try at trying to clean that code up in the past but didn't get very far before giving up myself.

As for how to compile and link `arm64-android` vcpkg code using `x86_64-pc-windows-msvc` rustup target, that doesn't seem right to me.  Wouldn't you have to be compiling with one of the targets output from `rustup target list | grep android`?  Something like `aarch64-linux-android` maybe I'm guessing would pair with `arm64-android`?

For what it is worth, vcpkg does *not* seem to offer all triplets working on all hosts.  For instance, many Windows libraries are just not possible from Linux afaict (maybe with some wine tooling and using shoehorning in MSFTs tools?), with many of even the mingw builds also just broken. I've hit this a few times where the vcpkg port really wants e.g. Powershell as a dev dependency when targeting Windows, failing on Linux hosts. YMMV.

**Comment by JohnScience** [comment]
Clarification: `x86_64-pc-windows-msvc` was not the rustup target but the host.

**Comment by kartonrad** [comment]
I'm having a similar issue
I want to build for x86_64-linux-android, from x86_64-pc-windows-msvc

And i have ffmpeg:x86-android installed with vcpkg.
However, vcpkg-rs refuses to discover the package:  `the vcpkg-rs Vcpkg build helper can only find libraries built for the MSVC ABI.`
